### PR TITLE
[BEAM-3713] Move PerformanceTest and CrossLanguageValidateRunner from nosetest to pytest

### PR DESCRIPTION
**Pull request by benWize**
Move integration tests from nose to pytest on the following jobs:

- job_PerformanceTests_Python
- job_PostCommit_CrossLanguageValidatesRunner_Dataflow

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by benWize** [comment]
Run XVR_Dataflow PostCommit

**Comment by benWize** [comment]
Run Python37 WordCountIT Performance Test

**Comment by benWize** [comment]
Run XVR_Dataflow PostCommit

**Comment by aaltay** [comment]
What is the next step on this PR?

**Comment by benWize** [comment]
Hi @aaltay! 
Some of these tests failed on pytest, I need to investigate further why it is happening. I'll retake this.

**Comment by benWize** [comment]
Run XVR_Dataflow PostCommit

**Comment by benWize** [comment]
Run XVR_Dataflow PostCommit

**Comment by benWize** [comment]
R: @udim, Hi Udi, would you help me to review this? 
I left the nosetests prefix to be able to see the JUnit file before merging.
There is a successful run of [PostCommit_CrossLanguageValidatesRunner_Dataflow](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow_PR/22/)
I think in order to test PerformanceTests_Python it'll be necessary to run a seed job because the test reference is in the Jenkins file https://github.com/apache/beam/blob/1caeef33bccb31f17f041b3666b8bcb0f7930a86/.test-infra/jenkins/job_PerformanceTests_Python.groovy#L40

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 1]
```diff
@@ -2101,13 +2101,13 @@ class BeamModulePlugin implements Plugin<Project> {
...
-          "--attr=${config.pythonTestAttr}"
-        ]
         def beamPythonTestPipelineOptions = [
           "pipeline_opts": config.pythonPipelineOptions + sdkLocationOpt,
-          "test_opts": testOpts + config.nosetestsOptions,
-          "suite": "xlangValidateRunner"
+          "test_opts": config.pytestOptions,
+          "suite": "xlangValidateRunner",
+          "pytest": true, // TODO(BEAM-3713): Remove this once nose is removed.
+          // TODO(BEAM-10507): Remove no_xlang_transforms after fixing
```
Comment (udim) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:2109]: I think you can remove this TODO. We will always want to exclude tests marked with `no_xlang_transforms`.

File: sdks/python/apache_beam/io/external/xlang_parquetio_test.py [hunk 2]
```diff
@@ -36,7 +36,7 @@
 PARQUET_WRITE_URN = "beam:transforms:xlang:test:parquet_write"
 
 
-@attr('UsesCrossLanguageTransforms')
+@pytest.mark.xlang_transforms
```
Comment (udim) [inline sdks/python/apache_beam/io/external/xlang_parquetio_test.py:39]: Alternatively, you could remove the `no_xlang_transforms` marker and just comment out this line with the same TODO.

**Comment by udim** [comment]
Run BigQueryIO Read Performance Test Python

**Comment by udim** [comment]
The run of `PostCommit_CrossLanguageValidatesRunner_Dataflow` looks okay. I ran the seed job on my pr and started a random performance test: https://ci-beam.apache.org/job/beam_PerformanceTests_BiqQueryIO_Read_Python_PR/3/

**Comment by udim** [comment]
beam_PerformanceTests_WordCountIT_Py37

**Comment by udim** [comment]
Run Python37 WordCountIT Performance Test

**Comment by udim** [comment]
https://ci-beam.apache.org/job/beam_PerformanceTests_WordCountIT_Py37/2959/console failed.

**Comment by benWize** [comment]
> https://ci-beam.apache.org/job/beam_PerformanceTests_WordCountIT_Py37/2959/console failed. I think the pytest timeout of 600s might be the issue

I think the same, I increased the timeout

File: sdks/python/apache_beam/io/external/xlang_parquetio_test.py (hunk 2 above)
Comment (benWize) [reply sdks/python/apache_beam/io/external/xlang_parquetio_test.py]: I did this, I removed the `no_xlang_transforms` marker and commented out those tests

**Comment by benWize** [comment]
Run XVR_Dataflow PostCommit

File: sdks/python/apache_beam/io/external/xlang_parquetio_test.py (hunk 2 above)
Comment (benWize) [reply sdks/python/apache_beam/io/external/xlang_parquetio_test.py]: A successful run with changes in marker `no_xlang_transforms ` https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow_PR/23/testReport/

**Comment by tysonjh** [comment]
Run Python37 WordCountIT Performance Test

**Comment by udim** [comment]
Run Python37 WordCountIT Performance Test

**Comment by udim** [comment]
run was successful: https://ci-beam.apache.org/job/beam_PerformanceTests_WordCountIT_Py37/3002/

**Review by udim — APPROVED** [review]
Looks good, just needs to resolve the temporary TODO(BEAM-3713)s

**Comment by benWize** [comment]
Thanks, Udi, I removed the TODOs.

**Comment by udim** [comment]
Run Java PreCommit

**Comment by udim** [comment]
run python precommit

**Comment by udim** [comment]
Ignoring the precommit failures that seem unrelated

**Comment by chamikaramj** [comment]
Seems like this broke portable xvr tests.

For example,
https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/
https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/1361/console

ModuleNotFoundError: No module named 'pytest'

(Flink and Spark test suites are broken as well)

**Comment by benWize** [comment]
> Seems like this broke portable xvr tests.
> 
> For example,
> https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/
> https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/1361/console
> 
> ModuleNotFoundError: No module named 'pytest'
> 
> (Flink and Spark test suites are broken as well)

Fix in #15058

**Comment by tvalentyn** [comment]
This change might have started to cause test flakiness: https://issues.apache.org/jira/browse/BEAM-12515

**Comment by udim** [comment]
I don't see how this PR would affect how PipelineOptionsTest.test_display_data runs. It was running with pytest before this PR.

**Comment by benWize** [comment]
I don't think this PR is related to that flakiness, I was able to reproduce it locally in code before this PR was merge using this plugin https://github.com/dropbox/pytest-flakefinder as:
`pytest -o junit_suite_name=py38-cloud --junitxml=pytest_py38-cloud.xml --log-cli-level=DEBUG "apache_beam/options/pipeline_options_test.py::PipelineOptionsTest"  --flake-finder --flake-runs=10 -n 9`.
The problem could be related to xdist, maybe adding the marker `@pytest.mark.no_xdist` to test_display_data should fix the problem.

**Comment by benWize** [comment]
Possible fix in #15104

**Pull request merged**